src/libNode/MicroBlockPreProcessing.cpp

```diff
@@ -653,16 +645,40 @@ void Node::ProcessTransactionWhenShardBackup() {
 
   cv_TxnProcFinished.notify_all();
 
+  ReinstateMemPool(t_addrNonceTxnMap, gasLimitExceededTxnBuffer);
+}
+
+void Node::ReinstateMemPool(
+    const map<Address, map<uint64_t, Transaction>>& addrNonceTxnMap,
+    const vector<Transaction>& gasLimitExceededTxnBuffer) {
+  unique_lock<shared_timed_mutex> g(m_unconfirmedTxnsMutex);
+
   // Put remaining txns back in pool
-  for (const auto& kv : t_addrNonceTxnMap) {
+  for (const auto& kv : addrNonceTxnMap) {
     for (const auto& nonceTxn : kv.second) {
       t_createdTxns.insert(nonceTxn.second);
+      m_unconfirmedTxns.emplace(nonceTxn.second.GetTranID(),
+                                PoolTxnStatus::PRESENT_NONCE_HIGH);
     }
   }
 
   for (const auto& t : gasLimitExceededTxnBuffer) {
     t_createdTxns.insert(t);
+    m_unconfirmedTxns.emplace(t.GetTranID(),
+                              PoolTxnStatus::PRESENT_GAS_EXCEEDED);
+  }
+}
+
+PoolTxnStatus Node::IsTxnInMemPool(const TxnHash& txhash) const {
+  shared_lock<shared_timed_mutex> g(m_unconfirmedTxnsMutex, defer_lock);
+  // Try to lock for 100 ms
+  if (!g.try_lock_for(chrono::milliseconds(100))) {
+    return PoolTxnStatus::ERROR;
+  }
+  if (m_unconfirmedTxns.find(txhash) == m_unconfirmedTxns.end()) {
```

Comment:
To avoid redundant search, you can put the result of `find` in a variable and access that for the last `return` statement

Reply:
done